hermes/src/store/storage/local_storage.rs

```diff
@@ -0,0 +1,163 @@
+                            }
+                        }
+
+                        // Binary search returns Ok(idx) if the element is found at index idx or Err(idx) if it's not
+                        // found which idx is the index where the element should be inserted to keep the vector sorted.
+                        // Getting idx within any of the match arms will give us the index of the element that is
+                        // closest after or equal to the requested time.
+                        let idx = match key_cache
+                            .binary_search_by_key(&time, |record| record.time().publish_time)
+                        {
+                            Ok(idx) => idx,
+                            Err(idx) => idx,
+                        };
+
+                        // We are using `get` to handle out of bound idx. This happens if the
+                        // requested time is after the last element in the vector.
+                        key_cache.get(idx).cloned()
+                    }
+                }
+            }
+            None => None,
+        }
+    }
+}
+
+#[async_trait]
+impl Storage for LocalStorage {
+    /// Add a new db entry to the cache.
+    ///
+    /// This method keeps the backed store sorted for efficiency, and removes
+    /// the oldest record in the cache if the max_size is reached. Entries are
+    /// usually added in increasing order and likely to be inserted near the
+    /// end of the deque. The function is optimized for this specific case.
+    async fn store_message_states(&self, message_states: Vec<MessageState>) -> Result<()> {
+        for message_state in message_states {
+            let key = message_state.key();
+
+            let mut key_cache = self.message_cache.entry(key).or_insert_with(VecDeque::new);
+
+            key_cache.push_back(message_state);
+
+            // Shift the pushed record until it's in the right place.
+            let mut i = key_cache.len().saturating_sub(1);
+            while i > 0 && key_cache[i - 1].time() > key_cache[i].time() {
+                key_cache.swap(i - 1, i);
+                i -= 1;
+            }
+
+            // FIXME remove equal elements by key and time
+
+            // Remove the oldest record if the max size is reached.
+            if key_cache.len() > self.cache_size as usize {
+                key_cache.pop_front();
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn fetch_message_states(
+        &self,
+        ids: Vec<PriceIdentifier>,
+        request_time: RequestTime,
+        filter: MessageStateFilter,
+    ) -> Result<Vec<MessageState>> {
+        ids.into_iter()
+            .flat_map(|id| {
+                let request_time = request_time.clone();
+                let message_types: Vec<MessageType> = match filter {
+                    MessageStateFilter::All => MessageType::iter().collect(),
+                    MessageStateFilter::Only(t) => vec![t],
+                };
```

Comment:
MessageStateFilter seems a bit overkill. You could just pass in `message_types_to_find` and the caller could provide `MessageType::iter().collect()` if they wanted all of them.

Reply:
Maybe, I don't have strong feeling about it. I personally feel this way has a more explicit boundary and `MessageType::iter()` is not something obvious.